src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -0,0 +1,571 @@
+        . Stream.timeIndexed
+
+------------------------------------------------------------------------------
+-- SQL Joins
+------------------------------------------------------------------------------
+--
+-- Some references:
+-- * https://en.wikipedia.org/wiki/Relational_algebra
+-- * https://en.wikipedia.org/wiki/Join_(SQL)
+
+-- TODO: OrdSet/IntSet/hashmap based versions of these. With Eq only
+-- constraint, the best would be to use an Array with linear search. If the
+-- second stream is sorted we can also use a binary search, using Ord
+-- constraint or an ordering function.
+--
+-- For Storables we can cache the second stream into an unboxed array for
+-- possibly faster access/compact representation?
+--
+-- If we do not want to keep the stream in memory but always read it from the
+-- source (disk/network) every time we iterate through it then we can do that
+-- too by reading the stream every time, the stream must have immutable state
+-- in that case and the user is responsible for the behavior if the stream
+-- source changes during iterations. We can also use an Unfold instead of
+-- stream. We probably need a way to distinguish streams that can be read
+-- mutliple times without any interference (e.g. unfolding a stream using an
+-- immutable handle would work i.e. using pread/pwrite instead of maintianing
+-- an offset in the handle).
+
+-- XXX We can do this concurrently.
+--
+-- | This is the same as 'Streamly.Internal.Data.Unfold.outerProduct' but less
+-- efficient.
+--
+-- The second stream is evaluated multiple times. If the second stream is
+-- consume-once stream then it can be cached in an 'Data.Array.Array' before
+-- calling this function. Caching may also improve performance if the stream is
+-- expensive to evaluate.
+--
+-- Time: O(m x n)
+--
+-- /Internal/
+{-# INLINE crossJoin #-}
+crossJoin :: Monad (t m) => t m a -> t m b -> t m (a, b)
+crossJoin s1 s2 = do
+    -- XXX use concatMap instead?
+    a <- s1
+    b <- s2
+    return (a, b)
+
+-- XXX We can do this concurrently.
+-- XXX If the second stream is sorted and passed as an Array we could use
+-- binary search if we have an Ord instance or Ordering returning function. The
+-- time complexity would then become (m x log n).
+--
+-- | For all elements in @t m a@, for all elements in @t m b@ if @a@ and @b@
+-- are equal by the given equality pedicate then return the tuple (a, b).
+--
+-- The second stream is evaluated multiple times. If the stream is a
+-- consume-once stream then the caller should cache it in an 'Data.Array.Array'
+-- before calling this function. Caching may also improve performance if the
+-- stream is expensive to evaluate.
+--
+-- For space efficiency use the smaller stream as the second stream.
+--
+-- Space: O(n) assuming the second stream is cached in memory.
+--
+-- Time: O(m x n)
+--
+-- /Internal/
+{-# INLINE innerJoin #-}
+innerJoin ::
+    forall (t :: (Type -> Type) -> Type -> Type) m a b.
```

Comment:
Why add this?

Reply:
IIRC - otherwise the code using it does not compile, compiler assumes the kind of `m` to be `Type`.